### ISSUE TITLE
routinator dump not working

### ISSUE THREAD

**Issue by cli0** [NLnetLabs/routinator#914]
I am running Routinator 0.13.0 (OS Ubuntu 20.04) and when I try to dump the routinator data I get the following error. The software creates the folder `cache`, creates the folder of `ta`-s inside of it but fails to create everything else. Even if I try to manually add `store/repositories.json` it won't recognize it.
```

$ routinator dump --output cache
[ERROR] Fatal: failed to write file /../cache/store/repositories.json: No such file or directory (os error 2)
[ERROR] Fatal error. Exiting.
```
Am I using the command wrong?

**Comment by partim** [comment]
No, there is definitely something broken. Manually creating the file doesn’t help because it deletes the target directory before starting to write out file so that there will be no old files that aren’t actually part of the store.

This error can only happen when the store is empty which should not happen if a validation run previously succeeded and produced at least one VRP.

The file in your error message (`/../cache/store/repositories.json`) looks a bit weird, though. It is supposed to be the absolute path it is trying to access and this is not really a valid path. Which could explain why it thinks the store is empty if it shouldn’t be. Maybe it just can’t find it?

**Comment by cli0** [comment]
Hello, thank you for the reply.

I made sure that routinator is running and `localhost:8323/json` produces VRPs. The command gives me the same error. It creates a folder called `cache/`, per the command, inside that folder it creates`ta/` with all the TALs and then apparently it tries to create the file `store/repositories.json` inside `cache\` and that's where it fails. The folder `store\` as well is not created.

**Comment by cli0** [comment]
hello guys, sorry to bother but is there an ETA for fixing the dump bug?

**Comment by partim** [comment]
That said, this is still strange: The error can only happen if the store is empty, which shouldn’t be the case and might suggest that you are using the wrong configuration. Our systemd units explicitely use the config from `/etc/routinator/routinator.conf` which is typcially set up to use a directory under `/var` somewhere while if you run it on command line it uses `~/.routinator.conf` and defaults to `~/.rpki-cache/repository` for the repository cache.

Can you check that your command line run is using the repository directory? Just run `routinator config` – it’ll be the first line of output.

**Comment by cli0** [comment]
Hi Martin,

the config line you asked for is `repository-dir = "/home/cli0/.rpki-cache/repository"` this is the default config value after a fresh install. No manual changes were made.

**Comment by partim** [comment]
Is that the directory used by the Routinator serving `localhost:8323`? If you’ve installed Routinator from our packages, that is likely `/var/lib/routinator/rpki-cache".

Asked differently: What happens if you run `routinator -c /etc/routinator/routinator.conf dump -o cache`?

**Comment by cli0** [comment]
I tested your command and the error persists, albeit it looks a little different now.

the content of my `routinator.conf` is
```
repository-dir = "/var/lib/routinator/rpki-cache"
rtr-listen = ["127.0.0.1:3323"]
http-listen = ["127.0.0.1:8323"]
```.

and the result of `routinator -c /etc/routinator/routinator.conf dump -o cache` is still
```
[ERROR] Fatal: failed to open directory /var/lib/routinator/rpki-cache/stored/rsync: No such file or directory (os error 2)
[ERROR] Fatal error. Exiting.
```
rsyncd is active and running, but in fact `/var/lib/routinator/rpki-cache/stored/` does not contain folder `rsync/`

**Comment by partim** [comment]
Aha! That’s more like it ;)

I can indeed reproduce this error. As a quick workaround, creating that missing directory should make the dump complete successfully.

I’ll go ahead with a release 0.13.1 in any case.

**Comment by cli0** [comment]
Indeed, it works with this workaround. Thank you!

I wonder though why `routinator config` gives a different `rpki-cache/` folder location than what is included in the .conf file. This is weird.

**Comment by partim** [comment]
If you look at the systemd unit file, you’ll find that its Routinator is started with an explicit config:

```
[Service]
ExecStart=/usr/bin/routinator --config=/etc/routinator/routinator.conf --syslog server
```

When started from the command line, it uses the per-user config or default values which use the user’s home directory.

This keeps causing issues, so we’ll rethink the approach for 0.14.

**Comment by cli0** [comment]
One last thing, since we are talking about configuration issues. Sometime last year after a routinator instance of mine was updated with the newest version, I noticed my customized config file was overwritten with the basic routinator config file. I don't know if this issue still persists but if it does it would be nice if routinator software updates did not silently rewrite the existing config files too.
